patchwork-registries/src/main/java/net/minecraftforge/registries/RegistryManager.java

```diff
@@ -43,28 +53,121 @@ public String getName() {
 
 	@SuppressWarnings("unchecked")
 	public <V extends IForgeRegistryEntry<V>> Class<V> getSuperType(Identifier key) {
-		return (Class<V>) RegistryClassMapping.getClass(key);
+		return (Class<V>) this.superTypes.inverse().get(key);
 	}
 
 	@SuppressWarnings("unchecked")
 	public <V extends IForgeRegistryEntry<V>> ForgeRegistry<V> getRegistry(Identifier key) {
-		return (ForgeRegistry<V>) this.registries.get(key);
+		Registry<?> vanillaRegistry = Registry.REGISTRIES.get(key);
+
+		if (vanillaRegistry instanceof ForgeRegistryProvider) {
+			return ((ForgeRegistryProvider) vanillaRegistry).getForgeRegistry();
+		}
+
+		return null;
 	}
 
 	public <V extends IForgeRegistryEntry<V>> IForgeRegistry<V> getRegistry(Class<? super V> clazz) {
-		return getRegistry(RegistryClassMapping.getIdentifier(clazz));
+		Identifier existingKey = this.getName(clazz);
+
+		if (existingKey == null) {
+			return null;
+		}
+
+		return getRegistry(existingKey);
 	}
 
 	public <V extends IForgeRegistryEntry<V>> Identifier getName(IForgeRegistry<V> reg) {
 		return reg.getRegistryName();
 	}
 
-	/**
-	 * Used by {@link net.minecraftforge.registries.ForgeRegistries}.
-	 *
-	 * @param registry the registry to add to the mapping
-	 */
-	void addRegistry(Identifier key, ForgeRegistry registry) {
-		registries.put(key, registry);
+	public Identifier getName(Class<?> clazz) {
+		if (clazz == null) {
+			return null;
+		}
+
+		do {
+			Identifier existingKey = RegistryManager.ACTIVE.superTypes.get(clazz);
+
+			if (existingKey != null) {
+				return existingKey;
+			}
+
+			clazz = clazz.getSuperclass();
+		} while (clazz != null && clazz != Object.class);
+
+		return null;
+	}
+
+	public Set<Identifier> getRegistryNames() {
+		return Registry.REGISTRIES.getIds();
+	}
+
+	<V extends IForgeRegistryEntry<V>> ForgeRegistry<V> createRegistry(Identifier name, RegistryBuilder<V> builder) {
+		Set<Class<?>> parents = Sets.newHashSet();
```

Comment:
This is a tiny nitpick, but please replace Sets.newHashSet() in this file with the diamond operator `new HashSet<>()`

Reply:
fixed